runtime/src/iree/base/queue.h

```diff
@@ -0,0 +1,88 @@
+// Copyright 2024 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef IREE_BASE_QUEUE_H__
+#define IREE_BASE_QUEUE_H__
+
+#include "iree/base/allocator.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+// The queue essentially a circular array
+// where we can push to the back and pop from the front.
+// The helper functions allow you to index into the array.
+// Furthermore an initial allocation may be provided inline
+// as an optimization.
+typedef struct iree_queue_t {
+  iree_allocator_t allocator;
+  uint8_t* elements;
+  iree_host_size_t element_size;
+  iree_host_size_t element_count;
+  iree_host_size_t capacity;
+  iree_host_size_t head;
+  uint8_t initial_allocation[];
+} iree_queue_t;
+
+// Initializes the queue with the given allocator, element size and
```

Comment:
```suggestion
// Initializes the queue with elements of the given |element_size|.
// Optionally |inline_count| can be provided to blahlblah.
```
(keeping the first line short is good as that's most commonly displayed in editors/code search/etc - no need to mention everything, just what's useful to a caller to know)